docs/modules/ROOT/pages/gsn.adoc

```diff
@@ -0,0 +1,77 @@
+
+==== acceptRelayedCall
+
+First, RelayHub will ask your recipient contract if it wants to receive a relayed call. Recall that you will be charged for incurred gas costs by the relay, so you should only accept calls that you're willing to pay for!
+
+There are multiple ways to make this work, including:
+ * having a whitelist of trusted users
+ * only accepting calls to an onboarding function
+ * charging users in tokens (possibly issued by you)
+ * delegating the acceptance logic off-chain
+
+All relayed call requests can be rejected at no cost to the recipient.
+
+=== pre and postRelayedCall
+
+You may have noticed one of the suggested acceptance flows involves charging a user in tokens. You may, for example, setup a system where users spend a special token to pay for relayed calls, and then either gift these to new users, or sell them off-chain. The GSN will let you easily implement these designs.
+
+After a relayed call is accepted, RelayHub will give your contract two opportunities to charge your user for their call, perform some bookeeping, etc.: _before_ and _after_ the actual relayed call is made. These functions are aptly named preRelayedCall and postRelayedCall.
+
+preRelayedCall will inform you of the maximum cost the call may have, and is used to charge the user in advance. This is only required in scenarios where your users may run off with their tokens before you're able to charge them, and not all applications need this function.
+
+postRelayedCall will give you an accurate estimate of the transaction cost, making it a natural place to charge users. It will also let you know if the relayed call reverted or not, which allows for more sophisticated interactions, like not charging users for reverted calls.
+
+== msg.sender and msg.data
+
+There's only one extra care you should take when working with GSN recipient contracts: _you must never use msg.sender or msg.data directly_. Understanding why requires an in-depth analysis of the inner workings of the GSN and RelayHub, but you need not worry about these low-level details: GSNRecipient provides two functions, `_msgSender()` and `_msgData()`, which are drop-in replacements for `msg.sender` and `msg.data`. As long as you use these two functions instead, you're good to go!
+
+== Payment
+
+By now you may be wondering how exactly relays charge their recipients for gas costs and service fee. The answer is simple: the recipient must deposit funds on RelayHub in advance, and payment is automatically handled on each relayed call.
+
+You can head to the [GSN dashboard](deltacamp.website) to check and top-up your contracts' balance, view previous charges, or you can do this programatically by calling IRelayHub.depositFor and IRelayHub.balanceOf.
+
+Recipients may withdraw their balance from the system at any point, but remember that they will not be able to receive any further relayed calls!
```

Comment:
Again, be careful with the subject: in the previous paragraph the recipient was "you", but on this one it's "they".

Reply:
This comes from the fact that it is the recipient itself (the contract) that must call `withdraw`, there's no notion of an 'owner' of a recipient, and I didn't want to mislead the reader. 

Perhaps it'd be best to briefly mention this, i.e. "don't forget to add a way to call withdraw from your contract!".

Reply:
> This comes from the fact that it is the recipient itself (the contract) that must call withdraw,

Damn, I wasn't taking this into consideration. This kinda forces you to have an owner of the contract who can call into withdraw.

Let's rephrase this sentence so it's clear that it is the contract who needs to withdraw, and add something (maybe an internal `withdraw` function?) in the GSNRecipientBase if we don't have one.